src/widget.cpp

```diff
@@ -694,10 +706,44 @@ void TextInputWidget::SetText(const std::string &text)
 	if (this->text_changed) this->text_changed();
 }
 
+/**
+ * Grab or release the keyboard focus.
+ * @param focus Widget should be focused.
+ */
+void TextInputWidget::SetFocus(bool focus)
+{
+	this->has_focus = focus;
+}
+
+bool TextInputWidget::OnClick([[maybe_unused]] const Point32 &base, const Point16 &pos)
+{
+	this->has_focus = true;
+
+	/* Find clicked cursor pos. \todo We might be able to use a more efficient logarithmic search. */
+	int w;
+	int text_len = this->buffer.size();
+	int best_x = INT_MAX;
+	for (int p = text_len; p >= 0; --p) {
+		_video.GetTextSize(this->buffer.substr(0, p), &w, nullptr, false);
+		if (abs(w - pos.x) <= abs(best_x - pos.x)) {
+			best_x = w;
+			this->cursor_pos = p;
+		} else {
+			break;
+		}
+	}
+
+	return true;
+}
+
 bool TextInputWidget::OnKeyEvent(WmKeyCode key_code, WmKeyMod mod, const std::string &symbol)
 {
-	if (!this->visible) return false;
+	if (!this->visible || !this->has_focus) return false;
 	switch (key_code) {
+		case WMKC_CANCEL:
+			this->has_focus = false;
```

Comment:
I wonder if you should to that for ENTER as well. For windows that disappear after finishing it makes no difference, but if the window is not closed, any further key press will change the last confirmed text-box now?